sdk/src/main/java/io/customer/sdk/queue/type/QueueRunner.kt

```diff
@@ -0,0 +1,39 @@
+package io.customer.sdk.queue.type
+
+import io.customer.sdk.api.CustomerIOAPIHttpClient
+import io.customer.sdk.extensions.valueOfOrNull
+import io.customer.sdk.queue.taskdata.TrackEventQueueTaskData
+import io.customer.sdk.util.JsonAdapter
+
+interface QueueRunner {
+    suspend fun runTask(task: QueueTask): QueueRunTaskResult
+}
+
+internal class QueueRunnerImpl(
+    private val jsonAdapter: JsonAdapter,
+    private val cioHttpClient: CustomerIOAPIHttpClient
+) : QueueRunner {
+    override suspend fun runTask(task: QueueTask): QueueRunTaskResult {
+        when (valueOfOrNull<QueueTaskType>(task.type)) {
+            QueueTaskType.IdentifyProfile -> TODO() // return identifyProfile(task)
+            QueueTaskType.TrackEvent -> return trackEvent(task)
+            null -> {
+                // TODO where hooks will attempt to run queue task in other module.
+            }
+        }
+
+        return Result.success(Unit)
+    }
+
+//    private suspend fun identifyProfile(task: QueueTask): QueueRunTaskResult {
+//        val taskData: IdentifyProfileQueueTaskData = jsonAdapter.fromJson(task.data)
+//
+//        // TODO perform http request
+//    }
+
+    private suspend fun trackEvent(task: QueueTask): QueueRunTaskResult {
```

Comment:
A bloated `QueueRunner` would indeed be more difficult to maintain. 

However, I do not recommend having a separate class/object for each type of task (`TrackTaskRunner`, `IdentifyTaskRunner`) as that's how [my old implementation of background queue](https://github.com/levibostian/wendy-android) works and after building dozens of apps using it, the boilerplate is really bad. Especially because the function bodies for each class/object are really small ([2 lines of code each](https://github.com/customerio/customerio-android/blob/13109bd7e1924e9354c0a300e32850f1c0d0f646/sdk/src/main/java/io/customer/sdk/queue/type/QueueRunner.kt#L35-L37)). 

To give you an idea of my plans are for fully implementing the `QueueRunner`, check out this code from the iOS SDK. [This class](https://github.com/customerio/customerio-ios/blob/0ca06183f9c0094019e3e58d0963b5931909c348/Sources/Tracking/Background%20Queue/QueueRunner.swift) is `QueueRunner` from the Tracking module and [this class](https://github.com/customerio/customerio-ios/blob/develop/Sources/MessagingPush/Background%20Queue/QueueRunner.swift) is the `QueueRunner` from the messasing push module. 

Those 2 classes are quite simple and small as each function is very small. I find this easier to maintain then separate classes/objects for each type of task. Therefore, I recommend we stay with this current implementation in the PR.

Reply:
A follow-up question, where do the non-CIO (local SDK ones) tasks get to run/triggered from. 
For example, when the customer is identified (200 OK) then save that identifier inside local prefs. 

Would `QueueRunner` is responsible for doing that?

Reply:
Great timing. I just got that implemented in another branch yesterday. 

The queue is only responsible for performing the network requests *after* [the SDK has already performed tasks such as saving the identifier to local prefs](https://github.com/customerio/customerio-android/blob/7e6b1d6724fd199b6a721e05a4726e48d6c19089/sdk/src/main/java/io/customer/sdk/CustomerIOClient.kt#L28-L67). 

The queue runner tasks are all [small](https://github.com/customerio/customerio-android/blob/7e6b1d6724fd199b6a721e05a4726e48d6c19089/sdk/src/main/java/io/customer/sdk/queue/QueueRunner.kt#L39-L43). Not necessarily on purpose, but it just works out that way because you want to do all the local changes to the state of the SDK before running the task in the queue. 

The queue acts as a sync to sync *snapshots of events* that have happened in the SDK. For example, let's say that you have an app that allows you to modify the first name of your profile. Your app wants to send your edited first name to the remote API of your app. Your app should do the following events:
1. Save the first name to local storage such as a database to act as a cache. 
2. Add a task to the background queue to eventually send the new first name to the remote API. It is at this time that you provide to the background queue the new first name to be sent to the remote API. It's like you are taking a snapshot of the state of the SDK at this time and giving it to the background queue. 

Why this is important is to avoid a scenario like this:
* Your first name in app is Dana. 
* You change your first name to Eddie. Add task to background queue to update first name. 
* Background queue task has not yet run yet. you change your first name to Frankie. Add task to background queue to update first name. 
* Background queue now has 2 tasks in it. The queue runs task 1 to update first name. If you provided the name "Eddie" to the background queue at the time of adding the event to the queue, the queue at this point would know to send "Eddie" to the API. But if you didn't provide "Eddie" and you are instead relying on your app simply checking the current state of the app's cached data, the background queue would instead send "Frankie" to the API and would never send Eddie. Your remote API would never know about Eddie, ever like it never happened. 

For some use cases like the one above, this should never be a problem. But what if you have a different use case that involves deleting? Such as a Twitter app allowing you to send tweets. If the following events happen in your app:
* Phone in airplane mode. User creates a new tweet. Task added to background queue to send tweet in the local database with ID 5 to remote API. 
* Phone still in airplane mode and user decides to delete tweet. Task added to background queue to delete tweet. App deletes row in database with ID 5. 
* Phone not in airplane and background queue runs. Queue task sees no row in database for ID 5, decides not to run. 
* .....what do you do now? The state of the background queue is now out of sync with your app's events? 

These examples might seem like they have simple solutions to solve them. I have built apps that implemented these simple solutions in them! But as the app grew, it quickly got out of hand. Instead, if you use your background queue as a queue of snapshots of historical events of your app, the queue will always be in sync with the state of your app. No edge cases to ever worry about. 

Did this answer your question or were you asking a different question? 
This is good info to capture. I'll make a PR to add these to the background queue docs.